virtual_rainforest/core/variables.py

```diff
@@ -0,0 +1,93 @@
+"""Module for all variables."""
+from dataclasses import dataclass, field
+from typing import Any
+
+
+@dataclass
+class Variable:
+    """Base class for all variables."""
+
+    name: str = field(init=False)
+    descritpion: str = field(init=False)
+    units: str = field(init=False)
+
+    def __init_subclass__(cls, name: str, descritpion: str, units: str) -> None:
+        """Checks if the variable class is unique and registers it.
+
+        Args:
+            name (str): Name of the variable.
+            descritpion (str): Description of the variable.
+            units (str): Units to use for the variable across the project.
+
+        Raises:
+            ValueError: If a variable class with the same name is already registered.
+            ValueError: If a variable with the same name field is already registered.
+        """
+        cls.name = name
+        cls.descritpion = descritpion
+        cls.units = units
+
+        if cls.__name__ in _UNIQUE_VARIABLES_CLASSES:
+            raise ValueError(f"Variable class {cls.__name__} already registered.")
+
+        if cls.name in _UNIQUE_VARIABLES_NAMES:
+            raise ValueError(f"Variable name {cls.name} already registered.")
+
+        _UNIQUE_VARIABLES_CLASSES.append(cls.__name__)
+        _UNIQUE_VARIABLES_NAMES.append(cls.name)
+
+    def __new__(cls) -> "Variable":
+        """Creates a new variable of this type or returns an existing one.
+
+        Returns:
+            Variable: The new or existing variable of this type.
+        """
+        if cls.name not in VARIABLES_REGISTRY:
+            VARIABLES_REGISTRY[cls.name] = super().__new__(cls)
+
+        return VARIABLES_REGISTRY[cls.name]
+
+    def __setattr__(self, __name: str, __value: Any) -> None:
+        """Raises an error if a variable is changed.
+
+        While we could use `frozen=True` in the dataclass, that will only prevent
+        modifying the existing attributes, but will not prevent adding new ones.
+        Overwritting __setattr__ is the way to prevent that both. Using `slots=True`
+        does not play well with `__init_subclass__`.
+        """
+        raise TypeError("Variables are immutable.")
```

Comment:
We would like to create/add variables during the setup() and first update(), would this be possible with this function? More generally, when would this registration happen? I think a lot of variables are initialized in the first update call.

Reply:
I think we'd either define them in `core.variables` (which we might want to break up into smaller units somehow because it could get big) or in a `model.variables` file? We would have to be able to do something like `model.variables` to support additional models. Then you would import the variables you need from those locations.